crates/biome_analyze/CONTRIBUTING.md

```diff
@@ -103,54 +105,154 @@ New rules **must** be placed inside the `nursery` group. This group is meant as
+2. Let's have a look at the generated code in  `use_my_rule_name.rs`:
 
-   ```shell
-   just new-css-lintrule useMyRuleName
+
+   ```rust
+   ...
+   impl Rule for UseMyRuleName {
+       type Query = Ast<JsIdentifierBinding>;
+       type State = ();
+       type Signals = Option<Self::State>;
+       type Options = ();
+
+       fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+           ...
+       }
+
+       fn diagnostic(ctx: &RuleContext<Self>, _state: &Self::State) -> Option<RuleDiagnostic> {
+           ...
+       }
+   }
    ```
 
-1. The `Ast` query type allows you to query the CST of a program.
-1. The `State` type doesn't have to be used, so it can be considered optional. However, it has to be defined as `type State = ()`.
-1. Implement the `run` function:
+   We can observe a few details from this snippet:
+
+   - The **`Options`** type can be used to define [additional options](#rule-options) for your rule:
+
+     ```rust
+     type Options = ();
+     ```
+
+     Use `()` if your rule does not have additional options.
+
+   - The **`Query`** type defines the entities for which your your rule's `UseMyRuleName::run` function will be invoked:
+     ```rust
+     type Query = Ast<JsIdentifierBinding>;
+     ```
+
+     _→ The `Ast<>` query type for example allows you to query the AST/CST of a program._
+
+     _→ For more advanced use cases it is also possible to define [custom query types](#custom-syntax-tree-visitors)._
+
+   - The **`run`** function will be invoked for each match of `Query`.
+     It should return either `Some` to report a diagnostic for that match, or `None`.
+
+     _→ It is also possible to report multiple diagnostics for a single `Query` match._
+     _See [Multiple Signals](#multiple-signals) for instructions._
+
+     ```rust
+     type Signals = Option<Self::State>;
+
+     fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+       let matched_node = ctx.query();
+       ...
+     }
+     ```
+
+   - The **`diagnostic`** function will be invoked for each occurence returned by `run`, and turns these occurences into `RuleDiagnostic` instances that define the message(s) reported to the user.
+
+     ```rust
+     fn diagnostic(ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> { ... }
+     ```
 
-   This function is called every time the analyzer finds a match for the query specified by the rule, and may return zero or more "signals".
+   - The `State` type can be used to pass additional information for each occurence reported by `run` to the `diagnostic` function.
+     Use `()` if you don't need to pass additional information:
 
-1. Implement the `diagnostic` function. Follow the [pillars](#explain-a-rule-to-the-user) when writing the message of a diagnostic:
+     ```rust
+     type State = ();
+     ```
+
+3. Optional: Use `Options` to [define custom options](#rule-options) for your rule.
+   We'll leave it as `()` for now.
```

Comment:
This paragraph repeats what we already said before at line 151. We should avoid repetitions

Reply:
The repetition is intentional here, because the two occurences serve different purposes and accomodate two different reading styles:

1. "Here's how a rule implementation looks in general"
2. "Here's what you have to do step-by-step to implement your example rule"

I'm trying to avoid burying the "General Design" information within the step-by-step guide so it's not overlooked if you don't know it's there (because that's what happened to me: I'm usually scanning the document for the piece of information I'm looking for by jumping between the headings and other important-looking visual cues, and if it looks like a step-by-step tutorial with a low information density after reading one or two paragraphs, I skip ahead).